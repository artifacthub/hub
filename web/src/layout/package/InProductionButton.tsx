import classnames from 'classnames';
import isNull from 'lodash/isNull';
import isUndefined from 'lodash/isUndefined';
import { useContext, useRef, useState } from 'react';
import { FaRegCheckCircle, FaRegCircle } from 'react-icons/fa';
import { MdBusiness } from 'react-icons/md';
import { RiMedalLine } from 'react-icons/ri';
import { useLocation, useNavigate } from 'react-router';

import API from '../../api';
import { AppCtx, signOut } from '../../context/AppCtx';
import useOutsideClick from '../../hooks/useOutsideClick';
import { ErrorKind, Organization, Repository } from '../../types';
import alertDispatcher from '../../utils/alertDispatcher';
import { getRepoKindName } from '../../utils/repoKind';
import ElementWithTooltip from '../common/ElementWithTooltip';
import Image from '../common/Image';
import Loading from '../common/Loading';
import styles from './InProductionButton.module.css';

interface Props {
  normalizedName: string;
  repository: Repository;
}

const InProductionButton = (props: Props) => {
  const { ctx, dispatch } = useContext(AppCtx);
  const navigate = useNavigate();
  const location = useLocation();
  const [openStatus, setOpenStatus] = useState(false);
  const [organizations, setOrganizations] = useState<Organization[] | undefined | null>(undefined);
  const [isLoading, setIsLoading] = useState<boolean>(false);
  const [updatingStatus, setUpdatingStatus] = useState<string | null>(null);

  const ref = useRef(null);
  useOutsideClick([ref], openStatus, () => setOpenStatus(false));

  async function fetchOrganizations(visibleLoading: boolean = false) {
    if (isNull(ctx.user)) {
      setOrganizations(undefined);
    } else {
      try {
        if (visibleLoading) {
          setIsLoading(true);
        }
        setOrganizations(
          await API.getProductionUsage({
            packageName: props.normalizedName,
            repositoryKind: getRepoKindName(props.repository.kind)!,
            repositoryName: props.repository.name,
          })
        );
        setOpenStatus(true);

        if (visibleLoading) {
          setIsLoading(false);
        }
      } catch (err: any) {
        setOrganizations(null);
        setOpenStatus(true);

        if (visibleLoading) {
          setIsLoading(false);
        }
        if (err.kind !== ErrorKind.Unauthorized) {
          alertDispatcher.postAlert({
            type: 'danger',
            message:
              'Something went wrong checking if your organizations use this package in production, please try again later.',
          });
        } else {
          dispatch(signOut());
          navigate(`${location.pathname}?modal=login&redirect=${location.pathname}`);
        }
      }
    }
  }

  async function changeUsage(name: string, isActive: boolean) {
    try {
      setUpdatingStatus(name);
      if (isActive) {
        await API.deleteProductionUsage(
          {
            packageName: props.normalizedName,
            repositoryKind: getRepoKindName(props.repository.kind)!,
            repositoryName: props.repository.name,
          },
          name
        );
      } else {
        await API.addProductionUsage(
          {
            packageName: props.normalizedName,
            repositoryKind: getRepoKindName(props.repository.kind)!,
            repositoryName: props.repository.name,
          },
          name
        );
      }
      alertDispatcher.postAlert({
        type: 'info',
        message: "Your change was applied successfully. It'll be visible across the site in a few minutes",
      });
      setUpdatingStatus(null);
      // We don't need to get orgs after changing it due to we are closing the dropdown
      // and we get them again every time we open the dropdown
      setOpenStatus(false);
    } catch (err: any) {
      setUpdatingStatus(null);
      setOpenStatus(false);

      if (err.kind !== ErrorKind.Unauthorized) {
        alertDispatcher.postAlert({
          type: 'danger',
          message: `${
            isActive
              ? 'Something went wrong deleting the selected organization from the list of production users of this package'
              : 'Something went wrong adding the selected organization to the list of production users of this package'
          }, please try again later.`,
        });
      } else {
        dispatch(signOut());
        navigate(`${location.pathname}?modal=login&redirect=${location.pathname}`);
      }
    }
  }

  const isDisabled = isNull(ctx.user) || isUndefined(ctx.user);

  return (
    <div className="d-none d-lg-block position-relative ms-2">
      <ElementWithTooltip
        active
        tooltipClassName={styles.tooltip}
        element={
          <button
            className={classnames('btn btn-outline-primary p-0 position-relative lh-1 fs-5', styles.iconWrapper, {
              [`disabled ${styles.isDisabled}`]: isDisabled,
            })}
            type="button"
            onClick={() => {
              if (!isLoading || isDisabled) {
                fetchOrganizations(true);
              }
            }}
            aria-label="Open organizations menu"
            aria-expanded={openStatus}
          >
            <div className="d-flex align-items-center justify-content-center">
              {isLoading && (
                <div className={styles.loading}>
                  <Loading noWrapper spinnerClassName={`position-absolute start-0 ${styles.spinner}`} />
                </div>
              )}
              <RiMedalLine className={classnames({ 'text-muted': isDisabled })} />
            </div>
          </button>
        }
        tooltipMessage="You must be signed in to specify which of your organizations are using this package in production"
        visibleTooltip={isDisabled}
      />

      <div
        ref={ref}
        role="menu"
        className={classnames('dropdown-menu dropdown-menu-end p-0', styles.dropdown, {
          show: openStatus,
        })}
      >
        <div
          className={classnames('dropdown-arrow', styles.arrow, {
            [styles.darkArrow]: organizations && organizations.length > 0,
          })}
        />

        {isNull(organizations) || isUndefined(organizations) || organizations.length === 0 ? (
          <div className={`p-4 text-center ${styles.emptyListMsg}`}>
            Here you'll be able to specify which of the organizations you belong to are using this package in
            production.
          </div>
        ) : (
          <div>
            <div className={`p-3 border-bottom border-1 ${styles.title}`}>
              Select which of your organizations are using this package in production
            </div>
            <div className={`overflow-auto ${styles.buttonsWrapper}`}>
              {organizations!.map((org: Organization) => {
                const isActive = org.usedInProduction || false;
                const isUpdating = !isNull(updatingStatus) && updatingStatus === org.name;

                return (
                  <button
                    className={`${styles.dropdownItem} dropdownItem btn p-3 rounded-0 w-100`}
                    onClick={() => changeUsage(org.name, isActive)}
                    key={`subs_${org.name}`}
                    aria-label={
                      isActive
                        ? `Delete ${org.displayName || org.name} organization from package's production users list`
                        : `Add ${org.displayName || org.name} organization to package's production users list`
                    }
                  >
                    <div className="d-flex flex-row align-items-start w-100 justify-content-between">
                      <div className="me-3 position-relative">
                        <span className={classnames({ 'd-none': isUpdating })}>
                          {isActive ? <FaRegCheckCircle className="text-success" /> : <FaRegCircle />}
                        </span>
                        {isUpdating && (
                          <div className="position-absolute top-0">
                            <Loading
                              spinnerClassName={`position-absolute ${styles.updatingSpinner}`}
                              noWrapper
                              smallSize
                            />
                          </div>
                        )}
                      </div>
                      <div className={`d-flex flex-column flex-grow-1 ${styles.growWidth}`}>
                        <div className="d-flex flex-row align-items-center">
                          <div className="me-2">
                            <div
                              className={`d-flex align-items-center justify-content-center overflow-hidden ${styles.imageWrapper}`}
                            >
                              <Image
                                alt={org.displayName || org.name}
                                imageId={org.logoImageId}
                                className={`m-auto ${styles.image}`}
                                placeholderIcon={<MdBusiness className={styles.orgIcon} />}
                              />
                            </div>
                          </div>
                          <div className="h6 mb-0 mw-100 text-truncate">{org.displayName || org.name}</div>
                        </div>
                      </div>
                    </div>
                  </button>
                );
              })}
            </div>
          </div>
        )}
      </div>
    </div>
  );
};

export default InProductionButton;
