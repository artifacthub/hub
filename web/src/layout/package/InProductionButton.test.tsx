import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { mocked } from 'jest-mock';
import { BrowserRouter as Router } from 'react-router-dom';

import API from '../../api';
import { AppCtx } from '../../context/AppCtx';
import { ErrorKind, Organization } from '../../types';
import alertDispatcher from '../../utils/alertDispatcher';
import InProductionButton from './InProductionButton';
jest.mock('../../api');
jest.mock('../../utils/alertDispatcher');

const mockDispatch = jest.fn();

const defaultProps = {
  normalizedName: 'pkgName',
  repository: {
    repositoryId: '0acb228c-17ab-4e50-85e9-ffc7102ea423',
    kind: 0,
    name: 'stable',
    displayName: 'Stable',
    url: 'http://repoUrl.com',
    userAlias: 'user',
    verifiedPublisher: false,
    official: false,
  },
};

const mockCtx = {
  user: { alias: 'userAlias', email: 'jsmith@email.com', passwordSet: true },
  prefs: {
    controlPanel: {},
    search: { limit: 60 },
    theme: {
      configured: 'light',
      effective: 'light',
    },
    notifications: {
      lastDisplayedTime: null,
      enabled: true,
      displayed: [],
    },
  },
};

const mockNotSignedInCtx = {
  user: null,
  prefs: {
    controlPanel: {},
    search: { limit: 60 },
    theme: {
      configured: 'light',
      effective: 'light',
    },
    notifications: {
      lastDisplayedTime: null,
      enabled: true,
      displayed: [],
    },
  },
};

const mockUndefinedUserCtx = {
  user: undefined,
  prefs: {
    controlPanel: {},
    search: { limit: 60 },
    theme: {
      configured: 'light',
      effective: 'light',
    },
    notifications: {
      lastDisplayedTime: null,
      enabled: true,
      displayed: [],
    },
  },
};

const getMockProductionUsage = (fixtureId: string): Organization[] => {
  return require(`./__fixtures__/InProductionButton/${fixtureId}.json`) as Organization[];
};

describe('InProductionButton', () => {
  afterEach(() => {
    jest.resetAllMocks();
  });

  it('creates snapshot', async () => {
    const mockProductionUsage = getMockProductionUsage('1');
    mocked(API).getProductionUsage.mockResolvedValue(mockProductionUsage);

    const { asFragment } = render(
      <AppCtx.Provider value={{ ctx: mockCtx, dispatch: jest.fn() }}>
        <Router>
          <InProductionButton {...defaultProps} />
        </Router>
      </AppCtx.Provider>
    );

    await userEvent.click(screen.getByRole('button', { name: /Open organizations menu/ }));

    await waitFor(() => {
      expect(API.getProductionUsage).toHaveBeenCalledTimes(1);
    });

    expect(await screen.findAllByRole('button')).toHaveLength(4);
    expect(asFragment()).toMatchSnapshot();
  });

  describe('Render', () => {
    describe('when user is signed in', () => {
      it('renders component', async () => {
        const mockProductionUsage = getMockProductionUsage('2');
        mocked(API).getProductionUsage.mockResolvedValue(mockProductionUsage);

        render(
          <AppCtx.Provider value={{ ctx: mockCtx, dispatch: jest.fn() }}>
            <Router>
              <InProductionButton {...defaultProps} />
            </Router>
          </AppCtx.Provider>
        );

        const button = screen.getByRole('button', { name: /Open organizations menu/ });
        expect(button).toBeInTheDocument();
        expect(button).not.toBeDisabled();
        await userEvent.click(button);

        await waitFor(() => {
          expect(API.getProductionUsage).toHaveBeenCalledTimes(1);
          expect(API.getProductionUsage).toHaveBeenCalledWith({
            packageName: 'pkgName',
            repositoryKind: 'helm',
            repositoryName: 'stable',
          });
        });

        await waitFor(() => {
          expect(screen.queryByRole('status')).toBeNull();
        });

        expect(screen.getByRole('menu')).toBeInTheDocument();
        expect(
          screen.getByText('Select which of your organizations are using this package in production')
        ).toBeInTheDocument();
        expect(screen.getByText('Artifact Hub')).toBeInTheDocument();
        expect(screen.getByText('Org test')).toBeInTheDocument();
        expect(screen.getByText('test1')).toBeInTheDocument();

        expect(
          screen.getByRole('button', {
            name: "Delete Artifact Hub organization from package's production users list",
          })
        ).toBeInTheDocument();
        expect(
          screen.getByRole('button', {
            name: "Add Org test organization to package's production users list",
          })
        ).toBeInTheDocument();
        expect(
          screen.getByRole('button', {
            name: "Add test1 organization to package's production users list",
          })
        ).toBeInTheDocument();
      });

      it('renders empty menu when any orgs', async () => {
        mocked(API).getProductionUsage.mockResolvedValue([]);

        render(
          <AppCtx.Provider value={{ ctx: mockCtx, dispatch: jest.fn() }}>
            <Router>
              <InProductionButton {...defaultProps} />
            </Router>
          </AppCtx.Provider>
        );

        const button = screen.getByRole('button', { name: /Open organizations menu/ });
        expect(button).toBeInTheDocument();
        expect(button).not.toBeDisabled();
        await userEvent.click(button);

        await waitFor(() => {
          expect(API.getProductionUsage).toHaveBeenCalledTimes(1);
          expect(API.getProductionUsage).toHaveBeenCalledWith({
            packageName: 'pkgName',
            repositoryKind: 'helm',
            repositoryName: 'stable',
          });
        });

        expect(screen.getByRole('menu')).toBeInTheDocument();
        expect(
          screen.getByText(
            "Here you'll be able to specify which of the organizations you belong to are using this package in production."
          )
        ).toBeInTheDocument();
      });

      it("deletes org from pkg's production users list", async () => {
        const mockProductionUsage = getMockProductionUsage('3');
        mocked(API).getProductionUsage.mockResolvedValue(mockProductionUsage);
        mocked(API).deleteProductionUsage.mockResolvedValue(null);

        render(
          <AppCtx.Provider value={{ ctx: mockCtx, dispatch: jest.fn() }}>
            <Router>
              <InProductionButton {...defaultProps} />
            </Router>
          </AppCtx.Provider>
        );

        await userEvent.click(screen.getByRole('button', { name: /Open organizations menu/ }));

        await waitFor(() => {
          expect(API.getProductionUsage).toHaveBeenCalledTimes(1);
        });

        const btn = await screen.findByRole('button', {
          name: "Delete Artifact Hub organization from package's production users list",
        });
        expect(btn).toBeInTheDocument();
        await userEvent.click(btn);

        await waitFor(() => {
          expect(API.deleteProductionUsage).toHaveBeenCalledTimes(1);
          expect(API.deleteProductionUsage).toHaveBeenCalledWith(
            {
              packageName: 'pkgName',
              repositoryKind: 'helm',
              repositoryName: 'stable',
            },
            'artifact-hub'
          );
        });

        await waitFor(() => {
          expect(alertDispatcher.postAlert).toHaveBeenCalledTimes(1);
          expect(alertDispatcher.postAlert).toHaveBeenCalledWith({
            type: 'info',
            message: "Your change was applied successfully. It'll be visible across the site in a few minutes",
          });
        });

        await waitFor(() => {
          expect(screen.getByRole('menu')).not.toHaveClass('show');
        });
      });

      it("adds org to pkg's production users list", async () => {
        const mockProductionUsage = getMockProductionUsage('4');
        mocked(API).getProductionUsage.mockResolvedValue(mockProductionUsage);
        mocked(API).addProductionUsage.mockResolvedValue(null);

        render(
          <AppCtx.Provider value={{ ctx: mockCtx, dispatch: jest.fn() }}>
            <Router>
              <InProductionButton {...defaultProps} />
            </Router>
          </AppCtx.Provider>
        );

        await userEvent.click(screen.getByRole('button', { name: /Open organizations menu/ }));

        await waitFor(() => {
          expect(API.getProductionUsage).toHaveBeenCalledTimes(1);
        });

        const btn = await screen.findByRole('button', {
          name: "Add Org test organization to package's production users list",
        });
        expect(btn).toBeInTheDocument();
        await userEvent.click(btn);

        await waitFor(() => {
          expect(API.addProductionUsage).toHaveBeenCalledTimes(1);
          expect(API.addProductionUsage).toHaveBeenCalledWith(
            {
              packageName: 'pkgName',
              repositoryKind: 'helm',
              repositoryName: 'stable',
            },
            'test'
          );
        });

        await waitFor(() => {
          expect(alertDispatcher.postAlert).toHaveBeenCalledTimes(1);
          expect(alertDispatcher.postAlert).toHaveBeenCalledWith({
            type: 'info',
            message: "Your change was applied successfully. It'll be visible across the site in a few minutes",
          });
        });

        await waitFor(() => {
          expect(screen.getByRole('menu')).not.toHaveClass('show');
        });
      });
    });

    describe('when getProductionUsage fails', () => {
      it('with any error', async () => {
        mocked(API).getProductionUsage.mockRejectedValue({ kind: ErrorKind.Other });

        render(
          <AppCtx.Provider value={{ ctx: mockCtx, dispatch: jest.fn() }}>
            <Router>
              <InProductionButton {...defaultProps} />
            </Router>
          </AppCtx.Provider>
        );

        await userEvent.click(screen.getByRole('button', { name: /Open organizations menu/ }));

        await waitFor(() => {
          expect(API.getProductionUsage).toHaveBeenCalledTimes(1);
          expect(API.getProductionUsage).toHaveBeenCalledWith({
            packageName: 'pkgName',
            repositoryKind: 'helm',
            repositoryName: 'stable',
          });
        });

        await waitFor(() => {
          expect(alertDispatcher.postAlert).toHaveBeenCalledTimes(1);
          expect(alertDispatcher.postAlert).toHaveBeenCalledWith({
            type: 'danger',
            message:
              'Something went wrong checking if your organizations use this package in production, please try again later.',
          });
        });
      });

      it('when user is not authorized', async () => {
        mocked(API).getProductionUsage.mockRejectedValue({ kind: ErrorKind.Unauthorized });

        render(
          <AppCtx.Provider value={{ ctx: mockCtx, dispatch: mockDispatch }}>
            <Router>
              <InProductionButton {...defaultProps} />
            </Router>
          </AppCtx.Provider>
        );

        await userEvent.click(screen.getByRole('button', { name: /Open organizations menu/ }));

        await waitFor(() => {
          expect(API.getProductionUsage).toHaveBeenCalledTimes(1);
          expect(API.getProductionUsage).toHaveBeenCalledWith({
            packageName: 'pkgName',
            repositoryKind: 'helm',
            repositoryName: 'stable',
          });
        });

        await waitFor(() => {
          expect(mockDispatch).toHaveBeenCalledTimes(1);
          expect(mockDispatch).toHaveBeenCalledWith({ type: 'signOut' });
        });
      });
    });

    describe('when addProductionUsage fails', () => {
      it('with any error', async () => {
        const mockProductionUsage = getMockProductionUsage('5');
        mocked(API).getProductionUsage.mockResolvedValue(mockProductionUsage);
        mocked(API).addProductionUsage.mockRejectedValue({ kind: ErrorKind.Other });

        render(
          <AppCtx.Provider value={{ ctx: mockCtx, dispatch: jest.fn() }}>
            <Router>
              <InProductionButton {...defaultProps} />
            </Router>
          </AppCtx.Provider>
        );

        await userEvent.click(screen.getByRole('button', { name: /Open organizations menu/ }));

        await waitFor(() => {
          expect(API.getProductionUsage).toHaveBeenCalledTimes(1);
        });

        const btn = await screen.findByRole('button', {
          name: "Add Org test organization to package's production users list",
        });
        expect(btn).toBeInTheDocument();
        await userEvent.click(btn);

        await waitFor(() => {
          expect(API.addProductionUsage).toHaveBeenCalledTimes(1);
          expect(API.addProductionUsage).toHaveBeenCalledWith(
            {
              packageName: 'pkgName',
              repositoryKind: 'helm',
              repositoryName: 'stable',
            },
            'test'
          );
        });

        await waitFor(() => {
          expect(alertDispatcher.postAlert).toHaveBeenCalledTimes(1);
          expect(alertDispatcher.postAlert).toHaveBeenCalledWith({
            type: 'danger',
            message:
              'Something went wrong adding the selected organization to the list of production users of this package, please try again later.',
          });
        });

        await waitFor(() => {
          expect(screen.getByRole('menu')).not.toHaveClass('show');
        });
      });

      it('when user is not authorized', async () => {
        const mockProductionUsage = getMockProductionUsage('5');
        mocked(API).getProductionUsage.mockResolvedValue(mockProductionUsage);
        mocked(API).addProductionUsage.mockRejectedValue({ kind: ErrorKind.Unauthorized });

        render(
          <AppCtx.Provider value={{ ctx: mockCtx, dispatch: mockDispatch }}>
            <Router>
              <InProductionButton {...defaultProps} />
            </Router>
          </AppCtx.Provider>
        );

        await userEvent.click(screen.getByRole('button', { name: /Open organizations menu/ }));

        await waitFor(() => {
          expect(API.getProductionUsage).toHaveBeenCalledTimes(1);
        });

        const btn = await screen.findByRole('button', {
          name: "Add Org test organization to package's production users list",
        });
        expect(btn).toBeInTheDocument();
        await userEvent.click(btn);

        await waitFor(() => {
          expect(API.addProductionUsage).toHaveBeenCalledTimes(1);
          expect(API.addProductionUsage).toHaveBeenCalledWith(
            {
              packageName: 'pkgName',
              repositoryKind: 'helm',
              repositoryName: 'stable',
            },
            'test'
          );
        });

        await waitFor(() => {
          expect(mockDispatch).toHaveBeenCalledTimes(1);
          expect(mockDispatch).toHaveBeenCalledWith({ type: 'signOut' });
        });
      });
    });

    describe('when deleteProductionUsage fails', () => {
      it('with any error', async () => {
        const mockProductionUsage = getMockProductionUsage('6');
        mocked(API).getProductionUsage.mockResolvedValue(mockProductionUsage);
        mocked(API).deleteProductionUsage.mockRejectedValue({ kind: ErrorKind.Other });

        render(
          <AppCtx.Provider value={{ ctx: mockCtx, dispatch: jest.fn() }}>
            <Router>
              <InProductionButton {...defaultProps} />
            </Router>
          </AppCtx.Provider>
        );

        await userEvent.click(screen.getByRole('button', { name: /Open organizations menu/ }));

        await waitFor(() => {
          expect(API.getProductionUsage).toHaveBeenCalledTimes(1);
        });

        const btn = await screen.findByRole('button', {
          name: "Delete Artifact Hub organization from package's production users list",
        });
        expect(btn).toBeInTheDocument();
        await userEvent.click(btn);

        await waitFor(() => {
          expect(API.deleteProductionUsage).toHaveBeenCalledTimes(1);
          expect(API.deleteProductionUsage).toHaveBeenCalledWith(
            {
              packageName: 'pkgName',
              repositoryKind: 'helm',
              repositoryName: 'stable',
            },
            'artifact-hub'
          );
        });

        await waitFor(() => {
          expect(alertDispatcher.postAlert).toHaveBeenCalledTimes(1);
          expect(alertDispatcher.postAlert).toHaveBeenCalledWith({
            type: 'danger',
            message:
              'Something went wrong deleting the selected organization from the list of production users of this package, please try again later.',
          });
        });

        await waitFor(() => {
          expect(screen.getByRole('menu')).not.toHaveClass('show');
        });
      });

      it('when user is not authorized', async () => {
        const mockProductionUsage = getMockProductionUsage('6');
        mocked(API).getProductionUsage.mockResolvedValue(mockProductionUsage);
        mocked(API).deleteProductionUsage.mockRejectedValue({ kind: ErrorKind.Unauthorized });

        render(
          <AppCtx.Provider value={{ ctx: mockCtx, dispatch: mockDispatch }}>
            <Router>
              <InProductionButton {...defaultProps} />
            </Router>
          </AppCtx.Provider>
        );

        await userEvent.click(screen.getByRole('button', { name: /Open organizations menu/ }));

        await waitFor(() => {
          expect(API.getProductionUsage).toHaveBeenCalledTimes(1);
        });

        const btn = await screen.findByRole('button', {
          name: "Delete Artifact Hub organization from package's production users list",
        });
        expect(btn).toBeInTheDocument();
        await userEvent.click(btn);

        await waitFor(() => {
          expect(API.deleteProductionUsage).toHaveBeenCalledTimes(1);
          expect(API.deleteProductionUsage).toHaveBeenCalledWith(
            {
              packageName: 'pkgName',
              repositoryKind: 'helm',
              repositoryName: 'stable',
            },
            'artifact-hub'
          );
        });

        await waitFor(() => {
          expect(mockDispatch).toHaveBeenCalledTimes(1);
          expect(mockDispatch).toHaveBeenCalledWith({ type: 'signOut' });
        });
      });
    });

    describe('displays disabled button', () => {
      it('when user is not signed in', async () => {
        render(
          <AppCtx.Provider value={{ ctx: mockNotSignedInCtx, dispatch: jest.fn() }}>
            <Router>
              <InProductionButton {...defaultProps} />
            </Router>
          </AppCtx.Provider>
        );

        const btn = await screen.findByRole('button', {
          name: /Open organizations menu/,
        });
        expect(btn).toBeInTheDocument();
        expect(btn).toHaveClass('disabled');

        await userEvent.hover(btn);

        expect(await screen.findByRole('tooltip')).toBeInTheDocument();
        expect(
          screen.getByText(
            'You must be signed in to specify which of your organizations are using this package in production'
          )
        ).toBeInTheDocument();
      });

      it('when user is undefined', async () => {
        render(
          <AppCtx.Provider value={{ ctx: mockUndefinedUserCtx, dispatch: jest.fn() }}>
            <Router>
              <InProductionButton {...defaultProps} />
            </Router>
          </AppCtx.Provider>
        );

        const btn = await screen.findByRole('button', {
          name: /Open organizations menu/,
        });
        expect(btn).toBeInTheDocument();
        expect(btn).toHaveClass('disabled');
      });
    });
  });
});
