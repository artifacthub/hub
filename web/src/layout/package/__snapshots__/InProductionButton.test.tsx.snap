// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InProductionButton creates snapshot 1`] = `
<DocumentFragment>
  <div
    class="d-none d-lg-block position-relative ms-2"
  >
    <div
      class="position-relative undefined"
    >
      <div
        data-testid="elementWithTooltip"
      >
        <button
          aria-expanded="true"
          aria-label="Open organizations menu"
          class="btn btn-outline-primary p-0 position-relative lh-1 fs-5 iconWrapper"
          type="button"
        >
          <div
            class="d-flex align-items-center justify-content-center"
          >
            <svg
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M12 7.00002C16.4183 7.00002 20 10.5817 20 15C20 19.4183 16.4183 23 12 23C7.58172 23 4 19.4183 4 15C4 10.5817 7.58172 7.00002 12 7.00002ZM12 9.00002C8.68629 9.00002 6 11.6863 6 15C6 18.3137 8.68629 21 12 21C15.3137 21 18 18.3137 18 15C18 11.6863 15.3137 9.00002 12 9.00002ZM12 10.5L13.3225 13.1797L16.2798 13.6094L14.1399 15.6953L14.645 18.6406L12 17.25L9.35497 18.6406L9.86012 15.6953L7.72025 13.6094L10.6775 13.1797L12 10.5ZM18 2.00002V5.00002L16.6366 6.13758C15.5305 5.55773 14.3025 5.17887 13.0011 5.04951L13 1.99902L18 2.00002ZM11 1.99902L10.9997 5.04943C9.6984 5.17866 8.47046 5.55738 7.36441 6.13706L6 5.00002V2.00002L11 1.99902Z"
              />
            </svg>
          </div>
        </button>
      </div>
    </div>
    <div
      class="dropdown-menu dropdown-menu-end p-0 dropdown show"
      role="menu"
    >
      <div
        class="dropdown-arrow arrow darkArrow"
      />
      <div>
        <div
          class="p-3 border-bottom border-1 title"
        >
          Select which of your organizations are using this package in production
        </div>
        <div
          class="overflow-auto buttonsWrapper"
        >
          <button
            aria-label="Delete Artifact Hub organization from package's production users list"
            class="dropdownItem dropdownItem btn p-3 rounded-0 w-100"
          >
            <div
              class="d-flex flex-row align-items-start w-100 justify-content-between"
            >
              <div
                class="me-3 position-relative"
              >
                <span
                  class=""
                >
                  <svg
                    class="text-success"
                    fill="currentColor"
                    height="1em"
                    stroke="currentColor"
                    stroke-width="0"
                    viewBox="0 0 512 512"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M256 8C119.033 8 8 119.033 8 256s111.033 248 248 248 248-111.033 248-248S392.967 8 256 8zm0 48c110.532 0 200 89.451 200 200 0 110.532-89.451 200-200 200-110.532 0-200-89.451-200-200 0-110.532 89.451-200 200-200m140.204 130.267l-22.536-22.718c-4.667-4.705-12.265-4.736-16.97-.068L215.346 303.697l-59.792-60.277c-4.667-4.705-12.265-4.736-16.97-.069l-22.719 22.536c-4.705 4.667-4.736 12.265-.068 16.971l90.781 91.516c4.667 4.705 12.265 4.736 16.97.068l172.589-171.204c4.704-4.668 4.734-12.266.067-16.971z"
                    />
                  </svg>
                </span>
              </div>
              <div
                class="d-flex flex-column flex-grow-1 growWidth"
              >
                <div
                  class="d-flex flex-row align-items-center"
                >
                  <div
                    class="me-2"
                  >
                    <div
                      class="d-flex align-items-center justify-content-center overflow-hidden imageWrapper"
                    >
                      <img
                        alt="Artifact Hub"
                        aria-hidden="true"
                        class="m-auto image"
                        src="/image/29a1adeb-4184-4ac1-8727-7307e14dd3e1"
                        srcset="/image/29a1adeb-4184-4ac1-8727-7307e14dd3e1@1x 1x, /image/29a1adeb-4184-4ac1-8727-7307e14dd3e1@2x 2x, /image/29a1adeb-4184-4ac1-8727-7307e14dd3e1@3x 3x, /image/29a1adeb-4184-4ac1-8727-7307e14dd3e1@4x 4x"
                      />
                    </div>
                  </div>
                  <div
                    class="h6 mb-0 mw-100 text-truncate"
                  >
                    Artifact Hub
                  </div>
                </div>
              </div>
            </div>
          </button>
          <button
            aria-label="Add Org test organization to package's production users list"
            class="dropdownItem dropdownItem btn p-3 rounded-0 w-100"
          >
            <div
              class="d-flex flex-row align-items-start w-100 justify-content-between"
            >
              <div
                class="me-3 position-relative"
              >
                <span
                  class=""
                >
                  <svg
                    fill="currentColor"
                    height="1em"
                    stroke="currentColor"
                    stroke-width="0"
                    viewBox="0 0 512 512"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8zm0 448c-110.5 0-200-89.5-200-200S145.5 56 256 56s200 89.5 200 200-89.5 200-200 200z"
                    />
                  </svg>
                </span>
              </div>
              <div
                class="d-flex flex-column flex-grow-1 growWidth"
              >
                <div
                  class="d-flex flex-row align-items-center"
                >
                  <div
                    class="me-2"
                  >
                    <div
                      class="d-flex align-items-center justify-content-center overflow-hidden imageWrapper"
                    >
                      <img
                        alt="Org test"
                        aria-hidden="true"
                        class="m-auto image"
                        src="/image/e1414600-6bbf-48a2-bf2b-dd1089bae95c"
                        srcset="/image/e1414600-6bbf-48a2-bf2b-dd1089bae95c@1x 1x, /image/e1414600-6bbf-48a2-bf2b-dd1089bae95c@2x 2x, /image/e1414600-6bbf-48a2-bf2b-dd1089bae95c@3x 3x, /image/e1414600-6bbf-48a2-bf2b-dd1089bae95c@4x 4x"
                      />
                    </div>
                  </div>
                  <div
                    class="h6 mb-0 mw-100 text-truncate"
                  >
                    Org test
                  </div>
                </div>
              </div>
            </div>
          </button>
          <button
            aria-label="Add test1 organization to package's production users list"
            class="dropdownItem dropdownItem btn p-3 rounded-0 w-100"
          >
            <div
              class="d-flex flex-row align-items-start w-100 justify-content-between"
            >
              <div
                class="me-3 position-relative"
              >
                <span
                  class=""
                >
                  <svg
                    fill="currentColor"
                    height="1em"
                    stroke="currentColor"
                    stroke-width="0"
                    viewBox="0 0 512 512"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M256 8C119 8 8 119 8 256s111 248 248 248 248-111 248-248S393 8 256 8zm0 448c-110.5 0-200-89.5-200-200S145.5 56 256 56s200 89.5 200 200-89.5 200-200 200z"
                    />
                  </svg>
                </span>
              </div>
              <div
                class="d-flex flex-column flex-grow-1 growWidth"
              >
                <div
                  class="d-flex flex-row align-items-center"
                >
                  <div
                    class="me-2"
                  >
                    <div
                      class="d-flex align-items-center justify-content-center overflow-hidden imageWrapper"
                    >
                      <svg
                        class="orgIcon"
                        fill="currentColor"
                        height="1em"
                        stroke="currentColor"
                        stroke-width="0"
                        viewBox="0 0 24 24"
                        width="1em"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M0 0h24v24H0z"
                          fill="none"
                        />
                        <path
                          d="M12 7V3H2v18h20V7H12zM6 19H4v-2h2v2zm0-4H4v-2h2v2zm0-4H4V9h2v2zm0-4H4V5h2v2zm4 12H8v-2h2v2zm0-4H8v-2h2v2zm0-4H8V9h2v2zm0-4H8V5h2v2zm10 12h-8v-2h2v-2h-2v-2h2v-2h-2V9h8v10zm-2-8h-2v2h2v-2zm0 4h-2v2h2v-2z"
                        />
                      </svg>
                    </div>
                  </div>
                  <div
                    class="h6 mb-0 mw-100 text-truncate"
                  >
                    test1
                  </div>
                </div>
              </div>
            </div>
          </button>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
